fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,566 @@
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+
+    fn public(&'_ self) -> &'_ Self::PubKey {
+        &self.name
+    }
+
+    fn private(self) -> Self::PrivKey {
+        Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap()
+    }
+
+    #[cfg(feature = "copy_key")]
+    fn copy(&self) -> Self {
+        Secp256r1KeyPair {
+            name: self.name.clone(),
+            secret: Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap(),
+        }
+    }
+
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self {
+        let privkey = ExternalSecretKey::random(rng);
+        let pubkey = ExternalPublicKey::from(&privkey);
+
+        Secp256r1KeyPair {
+            name: Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            },
+            secret: Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::new(),
+            },
+        }
+    }
+}
+
+impl FromStr for Secp256r1KeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl Signer<Secp256r1Signature> for Secp256r1KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Secp256r1Signature, signature::Error> {
+        // Code copied from Sign.rs in k256@0.11.6
+
+        // Hash message
+        let z = FieldBytes::from(Sha256::digest(msg).digest);
+
+        // Private key as scalar
+        let x = U256::from_be_bytes(self.secret.privkey.as_nonzero_scalar().to_bytes().into());
+
+        // Generate k deterministically according to RFC6979
+        let k = rfc6979::generate_k::<sha2::Sha256, U256>(&x, &NistP256::ORDER, &z, &[]);
+        let k = Scalar::from(ScalarCore::<NistP256>::new(*k).unwrap());
+
+        if k.borrow().is_zero().into() {
+            return Err(signature::Error::new());
+        }
+
+        let z = Scalar::from_be_bytes_reduced(z);
+
+        // Compute scalar inversion of 𝑘
+        let k_inv = Option::<Scalar>::from(k.invert()).ok_or_else(signature::Error::new)?;
+
+        // Compute 𝑹 = 𝑘×𝑮
+        let big_r = (ProjectivePoint::GENERATOR * k.borrow()).to_affine();
+
+        // Lift x-coordinate of 𝑹 (element of base field) into a serialized big
+        // integer, then reduce it into an element of the scalar field
+        let r = Scalar::from_be_bytes_reduced(big_r.x());
+
+        let x = Scalar::from_be_bytes_reduced(x.to_be_byte_array());
+
+        // Compute 𝒔 as a signature over 𝒓 and 𝒛.
+        let s = k_inv * (z + (r * x));
+
+        if s.is_zero().into() {
+            return Err(signature::Error::new());
+        }
+
+        let sig = ExternalSignature::from_scalars(r, s)?;
+
+        // Note: This line is introduced here because big_r.y is a private field.
+        let y: Scalar = Scalar::from_be_bytes_reduced(*big_r.to_encoded_point(false).y().unwrap());
```

Comment:
Can this fail with a malformed R? maybe avoid unwrap just in case. ah it's signing, we don't really care, as it will never run on-chain + R is computed here.

Reply:
The `y()` method only returns `None` if the encoded point is compressed or compact, but we specifically ask for an uncompressed encoding, so the unwrap will never fail in this case. I will make the code bit more clear.